design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+
+## Handling Unique VM CloudInit and ConfigMap Volumes at Scale
+
+By default, any secrets or configMaps references in a VMConfig’s Volume section will be used directly as is, without any modification to the naming. This means if you specify a secret in a CloudInitNoCloud volume within a VMConfig, that every VM instance spawned from that VMConfig will get the exact same secret used for their cloud-init user data.
+
+This default behavior can be modified by setting the **AppendPostfixToSecretReferences** and **AppendPostfixToConfigMapReferences** booleans to true on the VMPool spec. When these booleans are enabled, references to secret and configMap names will have the VM’s sequential postfix appended to the secret and configmap name. This allows someone to pre-generate unique per VM secret and configMap data for a VMPool ahead of time in a way that will be predictably assigned to VMs within the VMPool.
+
+## Autohealing
+
+When managing VMs at large scale, it is useful to automate the recovery of VMs which continually fail to reach a ready state. This automatically fixes scenarios where a VM’s state has somehow been corrupted and needs to be completely refreshed.
+
+By setting `AutohealingStrategy=ReprovisionOnFailure` on the VMPool’s spec, VMs which reach an unexpected failure state will automatically be completely deleted (including persistent storage) and re-provisioned. This allows for auto recovery in application scenarios that can withstand such an action.
+
+
+## Throttling Parallel VM Creation/Update
+
+The VMPool controller should establish some default upper limits when it comes to how many VMs can be batch created and updated at a time. This is similar to the VMIReplicaSet controller's internal `BurstReplicas` global variable. By default VMIReplicaSets will only create at most 250 VMIs at a time, but this is configurable as a global setting. The VMPool controller should adopt the same pattern of a max upper limit of 250 VMs per a VMPool with a global configurable setting.
+
+This setting is primarily meant as a way of preventing the VMPool controller from creating an unintentionally cluster DoS.
+
+## VirtualMachinePool vs VirtualMachineInstanceReplicaSet
+
+VMPools are designed to manage stateful VM workloads while VMIReplicaSets are designed to manage stateless VMI workloads.
+
+A VMIReplicaSet makes sense for a user who wants to manage stateless VMI objects using similar patterns to how Kubernetes ReplicaSets manage pods.
+
+A VMPool makes sense for a user who wants to manage stateful VM objects using similar patterns found in Iaas public clouds, like AWS, GCP, Azure. In this way, a VMPool is more similar to an AWS AutoscalingGroup or a GCP ManagedInstanceGroup than it is to any core Kubernetes API.
+
```

Comment:
Can we define some metrics?
- The number of VMIs restarted in a Pool
- The number of VMIs detached from a Pool
- Pool VMI perf metrics (A Pool can add a label to the phaseTransitionTimes so we can associate per pool)